### PR TITLE
Theme.json: Add block support feature level selectors for blocks

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related: https://github.com/WordPress/gutenberg/pull/31366

## What?

Allows custom CSS selectors for individual block support features e.g. for border support. This means styles generated for individual block support features can be applied to different elements within a block e.g the block wrapper or another element that can't be targeted via the elements API.

## Why?

While attempting to add border support to the image block we hit the issue where we need to skip border support serialization so we can apply the borders only to the inner `img` element.

If we add the `img` element to the Elements API, we end up creating a disjoint between the block editor and theme.json/global styles. The problem lies in the fact that we can't style the `img` elements via Global Styles. This leads us to a situation where a theme author could style all image block's `img` elements but if a user wished to tweak that styling they would have to edit every individual image.

Adding a "feature level selector" to theme.json and global styles means that we can keep the usual theme.json shape (no elements API use) and also keep using the Global Styles UI as per normal. The difference is that the generated styles target new custom CSS selectors.

## How?

- Adds support for defining `__experimentalSelector` prop within each individual block support
- Leverages the new feature level `__experimentalSelector` values to generate scoped CSS selectors specific to that block support
- Generates the styles for each block support opting into this, under the appropriate selector instead of under the block's root selector
- Some minor spelling and typo fixes

**Note: PHP unit tests were a little tricky to implement given that WordPress is loaded prior to the tests running meaning that WordPress' theme.json generates block metadata before our single, feature-level selector test could register a block. As a means of still having a programmatic test for the new selectors, I've added a `tests_add_filter()` call to the bootstrap to get a suitable block registered in time for testing. Once the Image block adopts these new selectors via https://github.com/WordPress/gutenberg/pull/31366 we could remove this filter and test via the image block.**

**We can also remove the `tests_add_filter` prior to merging this PR if we wish to keep things cleaner in the interim.**

## Testing Instructions
1. Run `npm run test-unit-php`
2. Run `npm run test-unit packages/edit-site/src/components/global-styles/test/use-global-styles-output.js`

#### Manual Testing

The generated styles can be manually tested via https://github.com/WordPress/gutenberg/pull/31366 which has been based on this PR.

1. Checkout https://github.com/WordPress/gutenberg/pull/31366
2. Edit your theme.json to style image blocks with a border (See snipped below for example image border config)
3. Create a new post, add an image, confirm the image border has been applied.
4. Save the post and view on the front end.
5. Inspect the image and ensure styles have been generated for `.wp-block-image img, .wp-block-image .wp-block-image__crop-area`

```json
{
	...
	"styles": {
		"blocks": {
			"core/image": {
				"border": {
					"radius": "9999px",
					"color": "fuchsia",
					"style": "dashed",
					"width": "3px"
				},
			}
		}
	}
}
```

Alternatively;

1. Update a block's support config to include an `__experimentalSelector` at the feature level (not root `supports` config)
2. Style that block and block support feature via theme.json
3. Load the editor and confirm styles have been generated for the block matching the selector you set.

### CODE REVIEW

**Comment by ramonjd** [comment]
> WordPress is loaded prior to the tests running meaning that WordPress' theme.json generates block metadata before our single, feature-level selector test could register a block

Just checking, are we blocked from testing manually due to this loading sequence?

**Comment by aaronrobertshaw** [comment]
> Just checking, are we blocked from testing manually due to this loading sequence?

Sorry for the confusion @ramonjd 

You can test manually either via:

1. https://github.com/WordPress/gutenberg/pull/31366/ which leverages the new feature level selectors from this PR.
2. Or, manually updating a block's block.json supports to include an `__experimentalSelector` flag within an individual block support configuration and then inspecting the theme.json or global styles generated styles for the expected resulting selectors.

I'll update the PR description and testing instructions.

File: packages/edit-site/src/components/global-styles/utils.js
Comment (ramonjd) [inline packages/edit-site/src/components/global-styles/utils.js:276]: Not a biggie for this PR but someday it'd be nice to put this one somewhere where duotone and global styles could access. 

I say "nice" but it's hard one to categorize 🤔 

I see we export `useCustomSides` from '@wordpress/block-editor' and use it in the [edit-side package.](https://github.com/WordPress/gutenberg/blob/3f3c8df23c70a37b7ac4dddebc82030362133593/packages/edit-site/src/components/global-styles/dimensions-panel.js#L12), but do you think we could/should do something similar for `scopeSelector()`and add it to the block-editor package API?

On the surface it looks like a general styles utility that might one day live in the style engine.

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/global-styles/utils.js:276]: This was one I meant to come back to and find a proper home for. 

> On the surface it looks like a general styles utility that might one day live in the style engine.

It is most relevant to the generation of styles so the style engine is probably the best long-term candidate. I wasn't sure whether it would be ok to add it to the block-editor package API or if that would be committing us to a deprecation cycle later on if it gets rolled into the style engine.

I don't feel strongly about it so happy to proceed if you'd recommend that approach.

File: phpunit/class-wp-theme-json-test.php
Comment (andrewserong) [inline phpunit/class-wp-theme-json-test.php]: Tiny nit: Does this test still need to unregister the block? If so, should the `unregister` be moved to before the `assertEquals` so that if the test fails it doesn't potentially affect other tests?

File: packages/edit-site/src/components/global-styles/utils.js
Comment (ramonjd) [reply packages/edit-site/src/components/global-styles/utils.js:276]: > I don't feel strongly about it so happy to proceed if you'd recommend that approach.

Me either 😄 

Maybe one to defer since it's doesn't really deal with the main thrust of the PR 👍

Comment (andrewserong) [reply packages/edit-site/src/components/global-styles/utils.js:276]: Since the implementation here matches the PHP implementation in the global styles class, it sounds like keeping it in the global styles utils for the time being might be okay. We could then copy / refactor the method to be put elsewhere when we need it?

**Review by andrewserong — COMMENTED** [review]
This is testing pretty nicely for me @aaronrobertshaw. It's working well for me in global styles, and the individual block-level border overrides the global styles version correctly on the site front-end:

| Global styles | Individual block overriding global styles (this is a different gallery block to the previous screenshot) |
| --- | --- |
|  image | image |

Overall, I quite like the idea of the feature, too — it slots in nicely with the existing `__experimentalSelector` and I like how it concatenates the root level selector with the feature-level one, too. While it is adding extra complexity to the global styles class, the changeset is a bit smaller than I'd imagined, so from my perspective, I think it's a worthwhile addition. Also the filter-based approach for ensuring a test block is available in the test environment sounds like a reasonable approach to me — it could be useful for other tests, too, having a test block around.

The one issue I ran into is that the border didn't appear to be displaying for me at the block level within the post editor, however, it renders correctly on the front end of the site: :

| Post editor | Site front end |
| --- | --- |
| image | image |

I'm testing with TwentyTwentyTwo theme and via checking out #31366, so I wasn't sure if it's more to do with that PR?

File: phpunit/class-wp-theme-json-test.php
Comment (aaronrobertshaw) [reply phpunit/class-wp-theme-json-test.php]: Honestly, I'd missed this when moving the block registration to the `tests_add_filter`. I did wonder though if unregistering the block limits unexpected results for other tests or creates more variability. I think I'd lean towards removing this.

**Comment by aaronrobertshaw** [comment]
Thanks for taking the time to review this one @andrewserong 👍 

> The one issue I ran into is that the border didn't appear to be displaying for me at the block level within the post editor, however, it renders correctly on the front end of the site: :

This one is due to WP core adding generic `img { border: none; }` and `iframe, img { border: 0; }` styles to the `common.css` for wp-admin. This means the CSS provided by border block support to apply default border styles via a `:where` selector is being overridden for `img` elements. It's therefore also a theme independent issue.

If I select a border style in the Post Editor the border displays correctly. 


https://user-images.githubusercontent.com/60436221/177275852-de0be7d0-5fb1-477d-a02f-4be46c595e0b.mp4



As you noted the issue is more to do with #31366 rather than this one so I don't think it's a blocker here. I'll address this in #31366, even if it is by adding `img:where()` selectors to apply the fallback/default border style that way.

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (ramonjd) [inline lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Probably a nitpick:

The comment is pretty clear, but I'm thinking about whether `SUPPORT_FEATURE` describes the role of the whitelist to folks digging about in the code.

Would `BLOCK_SUPPORT_FEATURES` or even `BLOCK_SUPPORT_FEATURE_LEVEL_SELECTORS` add anything?

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (ramonjd) [inline lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: I'm not sure what I'm seeing is correct, but I can't get multiple features to work at the same time, and I suspect it's because the features are either being unset or overwritten here.

So the border styles won't show when I have color/spacing/typography defined. 

<details>

<summary>block.json supports example</summary>

```json
	"supports": {
		"anchor": true,
		"align": true,
		"color": {
			"__experimentalSkipSerialization": true,
			"gradients": true,
			"__experimentalDefaultControls": {
				"background": true,
				"text": true
			},
			"__experimentalSelector": "> table"
		},
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalLetterSpacing": true,
			"__experimentalTextTransform": true,
			"__experimentalDefaultControls": {
				"fontSize": true
			},
			"__experimentalSelector": "> table"
		},
		"__experimentalBorder": {
			"__experimentalSkipSerialization": true,
			"color": true,
			"style": true,
			"width": true,
			"__experimentalSelector": "> table",
			"__experimentalDefaultControls": {
				"color": true,
				"style": true,
				"width": true
			}
		},
		"__experimentalSelector": ".wp-block-table",
		"__experimentalStyle": {
			"spacing": {
				"margin": "0 0 1em 0"
			}
		}
	},

```

</details>

<details>

<summary>theme.json styles</summary>

```json
"styles": {
		"blocks": {
			"core/cover": {
				"border": {
					"color": "fuchsia",
					"style": "dashed",
					"width": "3px"
				},
				"color": {
					"text": "fuchsia",
					"background": "black"
				},
				"typography": {
					"fontSize": "1px"
				},
				"spacing": {
					"padding": "100px"
				}
			},
			"core/table": {
				"border": {
					"color": "fuchsia",
					"style": "dashed",
					"width": "3px"
				},
				"color": {
					"text": "fuchsia",
					"background": "black"
				},
				"typography": {
					"fontSize": "100px"
				},
				"spacing": {
					"padding": "100px"
				}
			}
		}
	},

```

</details>

The only way I can get it to work is:

```php
// Generate the feature's declarations only.
if ( isset( $feature_declarations[ $feature_selector ] ) ) {
    $feature_declarations[ $feature_selector ] = array_merge( $feature_declarations[ $feature_selector ], static::compute_style_properties( $feature, $settings, null, $this->theme_json ) );
} else {
    $feature_declarations[ $feature_selector ] = static::compute_style_properties( $feature, $settings, null, $this->theme_json );
}
```

But it might an indicator of a deeper issue. 

It could be related to the fact that I'm seeing duplicate rules in the HTML source? Here's what I'm seeing when testing the cover block:

```html

<style id='global-styles-inline-css'> \* lots of styles here*\ .wp-block-cover > .wp-block-cover__inner-container > p{padding: 100px;} \* lots of styles here*\ .
</style>


<style id='wp-block-cover-inline-css'>
    .wp-block-cover{font-size: 1px;}.wp-block-cover > .wp-block-cover__inner-container > p{padding: 100px;}
</style>

```

**Review by ramonjd — COMMENTED** [review]
This is a nice change. And compact too!

With one rule, say, border, the selectors are output as expected in the frontend and site/block editors. 

So I see `.wp-block-table > table .some_selector`, where `.wp-block-table > table` is the root `__experimentalSelector` and `.some_selector` is the feature selector.

I was wondering if there'd ever be a case for `&` like selectors, where we could conditionally apply only one feature's style to the container? Sounds a bit weird when I type it out, but was just curious.

With various feature rules added, I noticed that CSS rules are missing when I activate multiple features (color/border etc). See [comment](https://github.com/WordPress/gutenberg/pull/42087/files#r914387841)

**Comment by aaronrobertshaw** [comment]
Thanks for reviewing @ramonjd 👍 

> I was wondering if there'd ever be a case for & like selectors, where we could conditionally apply only one feature's style to the container? Sounds a bit weird when I type it out, but was just curious.

I did wonder about something similar as well. I'm not sure how valid the following hypothetical would be but might illustrate the potential for a future desire to modify the root `__experimentalSelector` instead of only using it to scope an inner one.

It could be possible that we'd have a block that has a few Block Styles. Maybe we'd like to apply the block support styles only if a certain block style hasn't been selected. Such as;

- The block has a root `__experimentalSelector` of `.wp-block-example`.
- It might have a block style that would add `.is-example-block-style`.
- We wish to only apply a given block support's styles if that block style wasn't selected. The styles should also still be applied it to the same element as the root selector. e.g. via something like `&:not(.is-example-block-style)`.
- We'd end up with a feature level selector like: `.wp-block-example:not(.is-example-block-style)`

In future iterations of these feature level selectors, I'd like to think we could extend it to perhaps check if the feature level selector starts with `&` and resolve that when scoping the selector. For the time being though, the current scoping should solve immediate needs and buy us more time to consider changes.

> With various feature rules added, I noticed that CSS rules are missing when I activate multiple features (color/border etc). See [comment](https://github.com/WordPress/gutenberg/pull/42087/files#r914387841)

Thanks for flagging this, I'll look into it.

**Comment by andrewserong** [comment]
I had another thought (and it isn't at all a blocker for this PR), that when it comes to rendering block supports at the individual block level for server-rendered blocks, we might need to add a conditional that switches between inline styles (if no experimental selector is used) and a style block — so we'd potentially call the style engine differently based on whether or not the current feature uses the experimental selector? I imagine that'd be a bit neater than attempting to inject a `style` attribute further down the DOM tree of the server-rendered block.

Just thought I'd mention it as a potential to-do list item for the future, or in case it affects plans for server-rendered blocks like Search 🤔 (Apologies if this has been discussed already!)

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Thanks for catching this issue @ramonjd. I'd clearly been caught up testing different supports targeting different inner elements.

I've fixed this by [leveraging your snippet](https://github.com/WordPress/gutenberg/pull/42087/commits/e02ea3591df2b01ac6d048690b66428e680aa354#diff-f28aa79997745a4a51f917012b45a809c9b45d55134b55379a718f5c751dddf4R517-R521), adding an [explanatory inline comment](https://github.com/WordPress/gutenberg/pull/42087/commits/e02ea3591df2b01ac6d048690b66428e680aa354#diff-f28aa79997745a4a51f917012b45a809c9b45d55134b55379a718f5c751dddf4R514-R516), and [expanding the unit tests](https://github.com/WordPress/gutenberg/pull/42087/commits/e02ea3591df2b01ac6d048690b66428e680aa354#diff-a51f75ea7a86ddd18edea3fe94a0482427795fcd7b247e6959149cbac60964dfR108-R120) to cover:

- two supports using a shared custom feature level selector e.g. `.inner`
- one other support using its own unique selector e.g. `.sub-heading`
- and, a further support that doesn't have a feature level selector specified.


> But it might an indicator of a deeper issue.
> 
> It could be related to the fact that I'm seeing duplicate rules in the HTML source? Here's what I'm seeing when testing the cover block:

This looks to be unrelated to me. There was definitely a flaw in the previous logic generating the feature styles.

**Comment by aaronrobertshaw** [comment]
> when it comes to rendering block supports at the individual block level for server-rendered blocks, we might need to add a conditional that switches between inline styles (if no experimental selector is used) and a style block

I see these selectors primarily being used when skipping the serialization of individual block styles. A means to offer control via both theme.json and Global Styles when the styles generated would otherwise be incorrectly targetted.

If skipping serialization the block would have the styles manually generated and applied directly where it sees fit during its render callback. Could the choice between inline styles or a style block be made here avoiding an additional config flag?

Do you think it makes sense that we only enable the feature level selectors if the block support also skips serialization?

> I imagine that'd be a bit neater than attempting to inject a style attribute further down the DOM tree of the server-rendered block.

It might do although I'm not sure if we'd then run into specificity issues. For all their downsides, the inline styles make it clear that the individual block's support styles will be taking precedence.

**Comment by andrewserong** [comment]
> For all their downsides, the inline styles make it clear that the individual block's support styles will be taking precedence.

Agreed! It's one of the things I like about the Image block PR — that it seems quite natural injecting the inline style at the correct level in the hierarchy in the JS implementation.

> Do you think it makes sense that we only enable the feature level selectors if the block support also skips serialization?

Yes, I think that's a good way to handle it at the individual block support level. It means that in the shorter-term, this is a feature for blocks that skip serialization, but it doesn't stop us from exploring options in the style engine further down the track, if we want to leverage the feature-level selector for outputting `style` block rules. Sounds good to me! (We probably don't need to make any changes in this PR, we could leave that to when we tackle a server-side block like Search, if it makes things simpler 🙂)

**Comment by aaronrobertshaw** [comment]
> We probably don't need to make any changes in this PR, we could leave that to when we tackle a server-side block like Search, if it makes things simpler

Sounds like a plan to me. Happy to tweak the approach as more blocks leverage feature-level selectors but for now it would be great to be able to land this to unblock the application of borders on images in https://github.com/WordPress/gutenberg/pull/31366. 

I'll add a few extra reviewers and 🤞 that we can get this merged in the near future.

**Comment by andrewserong** [comment]
> I'll add a few extra reviewers and 🤞 that we can get this merged in the near future.

Sounds great — I think this feature is looking really good, and the scope of this PR is nice and incremental, so opening up for wider review sounds like a plan to me, too!

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (ramonjd) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: > It could be related to the fact that I'm seeing duplicate rules in the HTML source? Here's what I'm seeing when testing the cover block:

Thanks for looking into it. 

I'm also seeing it on trunk and suspect it's because we're enqueuing block styles in both places after https://github.com/WordPress/gutenberg/pull/42005

Edit: fix attempt [here](https://github.com/WordPress/gutenberg/pull/42207)

**Comment by ajlende** [comment]
@aaronrobertshaw Nice to see someone else working on this! I did [an exploration in a gist](https://gist.github.com/ajlende/645f2e08aabff84d7498cfc9ef868f24#file-block-jsonc-L13) a while back for multiple selectors in the context of using them to to apply SVG filters (duotone) to the inner parts of a block since `filter` is also a non-inherited property like `border`.

I'd like to stabilize a method of allowing blocks to choose multiple selectors and wrap `__experimentalDuotone` and `__experimentalSelector` into one; although, to be clear, that doesn't need to happen right now in this PR. In stabilizing it, I wanted to make room for plugins to utilize the selectors that a block specifies by adding generic selectors like `media` (for thing like duotone) or `container` (for things like border) rather than tying things to specific block support features.

I'm in transit to a meetup at the moment, so I'll do a proper code review when I arrive, but I figured it would be worth sharing my initial thoughts first. If the details of stabilizing that API are too far out of scope for here, feel free to comment on the gist or I can open up a proper discussion when I have a bit more time.

**Comment by aaronrobertshaw** [comment]
Thanks for dropping the link to your Gist @ajlende, some very interesting explorations there! 👍 

After giving it a read-through I have a couple of immediate thoughts or concerns.
1. **Portability**: The introduction of preprocessor filters could really hamper one of the main goals of theme.json i.e. its portability of styles across themes or even via block patterns, template parts etc.
2. **Scope of change**:
	- **Blocking** - The shape changes to theme.json are pretty massive. They'll no doubt generate plenty of discussion and take considerable time to reach a consensus on how best to move those ideas forward. I see this as blocking progress on improvements that have already been waiting a long time ( e.g. image borders for over a year ).
	- **Assessing impact** - It was also raised in some internal discussions with @andrewserong / @ramonjd that the level of change would be very hard to determine what sorts of impacts it would have (and whether the effort to make the change would be worthwhile overall). That said, we're all keen to see where the discussion leads in time.
	  
I propose that we move forward with the smaller incremental change this PR represents assuming you cannot see any blockers. I don't believe it will introduce any barriers to reimagining theme.json that don't already exist. Not stabilising the `__experimentalSelector` API change for the moment should keep the window open for some future tweaking as well.

File: packages/edit-site/src/components/global-styles/utils.js
Comment (ajlende) [reply packages/edit-site/src/components/global-styles/utils.js:276]: > Not a biggie for this PR but someday it'd be nice to put this one somewhere where duotone and global styles could access.

It would be nice if we didn't need to use `scopeSelector` in duotone at all! The usage of `useInstanceId` in duotone makes it tricky, but it would be nice to lift that out of the `editor.BlockListBlock` filter.

The best place for it is where it is in this PR, in my opinion.

**Review by ajlende — APPROVED** [review]
Looking good! I'm happy where this is going!

**Comment by aaronrobertshaw** [comment]
I've rebased this and resolved the conflicts after https://github.com/WordPress/gutenberg/pull/40875 landed. The PR is still working for me and the unit tests pass. Any extra testing is still welcome of course.

@jorgefilipecosta do you have any thoughts or objections to the approach taken in this PR? If not, I'll look to merge this soon.

**Review by ramonjd — APPROVED** [review]
I've taken this for another spin and, despite trying, I can't break it 😄 

I've added `__experimentalSelector` values to different styles in block.json and then styled blocks via theme.json.

The correct CSS is generated even when I mix selectors and use multiple styles.

Great stuff.

**Review by andrewserong — APPROVED** [review]
Just thought I'd take this for a spin, too — the code change looks good to me, and it's working nicely in both the editor and global styles. Nice work honing in on this feature! 🚀 

image

**Comment by jorgefilipecosta** [comment]
> I've rebased this and resolved the conflicts after #40875 landed. The PR is still working for me and the unit tests pass. Any extra testing is still welcome of course.
> 
> @jorgefilipecosta do you have any thoughts or objections to the approach taken in this PR? If not, I'll look to merge this soon.

This PR seems to have been very well discussed and passed by many reviewers so I trust your judgment here.

**Comment by aaronrobertshaw** [comment]
Thanks everyone for the reviews, testing and feedback. Appreciate it 👍

**Comment by carolinan** [comment]
Do I understand correctly that I do not make any changes or add any new keys to theme.json? 
If the selector changes without me making changes to theme.json, wont that break existing styles?

**Comment by aaronrobertshaw** [comment]
Hi @carolinan 👋 

I'm not sure I follow your questions exactly, so my apologies if I miss the mark answering them here. 

> Do I understand correctly that I do not make any changes or add any new keys to theme.json?

No new keys or changes to your theme.json are required. A block support feature's values are configured the same way whether they are applied to the block's root wrapper or a specific selector for that feature. 

When a theme.json configuration is being processed, CSS declarations are created for each block. These are then converted to a ruleset for a given selector. Before this PR, the block could define its root level selector via `supports.__experimentalSelector` or fall back to the default block class, e.g. `wp-block-image`. 

Now a block can define the selector for a specific block support feature, e.g. `supports.__experimentalBorder.__experimentalSelector`.  If a custom selector is set for a specific block support feature, any CSS declarations for that support feature are converted to a ruleset using the custom selector instead of the block's root selector.

> If the selector changes without me making changes to theme.json, wont that break existing styles?

If I understand correctly, you are concerned a block might suddenly change the selector configured in its block.json? The same could happen with the root level `__experimentalSelector` as it does not actually have to target the root element for a block. 

If a block is changing its markup such that it needs a new selector to apply feature-level styles, it would likely already be needing a deprecation or to maintain the old selector to maintain stylings on deprecated blocks. The `__experimentalSelector` flags allow for comma-separated selectors, which could be an option in such a situation.

Hope that helps clarify things 🤞

**Comment by carolinan** [comment]
Yes thank you. It doesn't literally "Add a "feature level selector" to theme.json ".

**Comment by aaronrobertshaw** [comment]
> It doesn't literally "Add a "feature level selector" to theme.json ".

Correct. Apologies if the PR title is misleading.

> Theme.json: Add block support feature level selectors for blocks

It adds processing and support for feature level selectors to the theme.json class ( e.g. `WP_Theme_JSON_6_1` ) and global styles.

**Pull request merged**
